src/test/java/com/google/api/codegen/testdata/csharp_gapic_resourcenames_library.baseline

```diff
@@ -0,0 +1,685 @@
+
+        public static BookNameOneof From(ArchivedBookName archivedBookName) => new BookNameOneof(Type.ARCHIVED_BOOK_NAME, archivedBookName);
+
+        public static BookNameOneof From(DeletedBookNameFixed deletedBookNameFixed) => new BookNameOneof(Type.DELETED_BOOK_NAME_FIXED, deletedBookNameFixed);
+
+        public BookNameOneof(Type type, IResourceName value)
+        {
+            _type = type;
+            _value = GaxPreconditions.CheckNotNull(value, nameof(value));
+        }
+
+        private Type _type;
+        private IResourceName _value;
+
+        // Possibly move to base class (if we have one?)
+        private T CheckAndReturn<T>(Type type)
+        {
+            if (_type != type)
+            {
+                throw new InvalidOperationException($"Requested type {type}, but this one-of contains type {_type}");
+            }
+            return (T)_value;
+        }
+
+        public BookName BookName => CheckAndReturn<BookName>(Type.BOOK_NAME);
+
+        public ArchivedBookName ArchivedBookName => CheckAndReturn<ArchivedBookName>(Type.ARCHIVED_BOOK_NAME);
+
+        public DeletedBookNameFixed DeletedBookNameFixed => CheckAndReturn<DeletedBookNameFixed>(Type.DELETED_BOOK_NAME_FIXED);
+
+        public override string ToString() => _value.ToString();
+    }
+
+    public sealed partial class DeletedBookNameFixed : IResourceName
+    {
+        private const string FixedValue = "_deleted-book_";
+
+        public static DeletedBookNameFixed Instance => new DeletedBookNameFixed();
+
+        public static DeletedBookNameFixed Parse(string deletedBookNameFixed)
+        {
+            DeletedBookNameFixed result;
+            if (!TryParse(deletedBookNameFixed, out result))
+            {
+                throw new ArgumentException($"Invalid resource name, must be \"{FixedValue}\"", nameof(deletedBookNameFixed));
+            }
+            return result;
+        }
+
+        public static bool TryParse(string deletedBookNameFixed, out DeletedBookNameFixed result)
+        {
+            GaxPreconditions.CheckNotNull(deletedBookNameFixed, nameof(deletedBookNameFixed));
+            if (deletedBookNameFixed == FixedValue)
+            {
+                result = Instance;
+                return true;
+            }
+            else
+            {
+                result = null;
+                return false;
+            }
+        }
+
+        private DeletedBookNameFixed() { }
+
+        public override string ToString() => FixedValue;
+    }
+
+
+    public partial class FindRelatedBooksResponse
+    {
+        /// <summary>
+        /// A <see cref="IEnumerable<BookName>"/>-typed view over the <see cref="Names"/> resource-name property.
+        /// </summary>
+        public IEnumerable<BookName> NamesAsResources
+        {
+            get
+            {
+                return Names.Select(x => BookName.Parse(x));
+            }
+            set
```

Comment:
The introduction of a setter here (when there isn't one for the collection itself) feels a little unnatural.

It might be better to expose an `IList<T>` here instead, which just delegates to the underlying collection. We could create a new one each time the getter is called, or add a field (at the cost of memory) and use that all the time. Let's talk about this offline.